conformance/tests/exceptions_context_managers.py

```diff
@@ -0,0 +1,85 @@
+"""
+Tests the handling of __exit__ return types for context managers.
+"""
+
+# Specification: https://typing.readthedocs.io/en/latest/spec/exceptions.html
+
+
+from typing import Any, Literal
+
+
+class CMBase:
+    def __enter__(self) -> None:
+        pass
+
+
+class Suppress1(CMBase):
+    def __exit__(self, exc_type, exc_value, traceback) -> bool:
+        return True
+
+
+class Suppress2(CMBase):
+    def __exit__(self, exc_type, exc_value, traceback) -> Literal[True]:
+        return True
+
+
+class NoSuppress1(CMBase):
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        return None
+
+
+class NoSuppress2(CMBase):
+    def __exit__(self, exc_type, exc_value, traceback) -> Literal[False]:
+        return False
+
+
+class NoSuppress3(CMBase):
+    def __exit__(self, exc_type, exc_value, traceback) -> Any:
+        return False
+
+
+class NoSuppress4(CMBase):
+    def __exit__(self, exc_type, exc_value, traceback) -> None | bool:
+        return None
+
+
+def func1() -> None:
+    with Suppress1():
+        raise ValueError("This exception is suppressed")
+
+    return 1  # E
+
+
+def func2() -> None:
+    with Suppress2():
+        raise ValueError("This exception is suppressed")
+
+    return 1  # E
+
+
+def func3() -> None:
+    with NoSuppress1():
+        raise ValueError("This exception is not suppressed")
+
+    return 1  # OK
```

Comment:
I feel a type checker that recognizes that this code is unreachable could still reasonably show an error here, since the return type is definitely inconsistent with the annotation. Therefore, I'd prefer to use a different mechanism to test the desired semantics here.

This seems to work well:

```python
from typing import Any, Literal, assert_type


class CMBase:
    def __enter__(self) -> None:
        pass


class Suppress1(CMBase):
    def __exit__(self, exc_type, exc_value, traceback) -> bool:
        return True


class Suppress2(CMBase):
    def __exit__(self, exc_type, exc_value, traceback) -> Literal[True]:
        return True


class NoSuppress1(CMBase):
    def __exit__(self, exc_type, exc_value, traceback) -> None:
        return None


def suppress1(x: int | str) -> None:
    if isinstance(x, int):
        with Suppress1():
            raise ValueError
    assert_type(x, int | str)


def nosuppress1(x: int | str) -> None:
    if isinstance(x, int):
        with NoSuppress1():
            raise ValueError
    assert_type(x, str)
```

[pyright](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFADJIwCmIAhgDZl0DOLtMA%2BvAtQLAAoQYIDGDViygBhALIAhVtQBcgqGqgATasCidO1FDRB6AFOwbAAlFAC0APigA5MCmWr1HhBOFCBYiVAAygCuCAgg1GwAjCayCuyWKgIeWjp61AAeVKbmwGSZItyI1PkZhQBujMElsPQi1ABGdCIA1tb2UA1gYAxJHh4RMMEgKFAAKiDVPqLibEGh4ZEsAEyx8oqJ7mqpuvpZXJxm1BalhTw1BZyVDNVkMHWNzW22DpRGjADaE9UAun39akGw1G334vhmAWcITCEWia3i1E2yXUO3S%2BxyxzyUEu51OVyqNXuzUerXaDmcrn%2BAKBIycLjBPh2LAWsJYMQySkwhigAB8oCx7mS6ZStpgdEgWKgBXQUPUTBkyKgYEiAWoAO5UAAW8xhSxiKtVAzoEuoUAAagSAKIgcAgUUSDhFXjyxXcvkCkCWaYCHYoMDM3Vwjlc-DuwUvYVuZFqJDiyUoaWy6gukMGgEamDaqEsvUmNOGqD0E3mq023D2tiO84pj1egRAA), [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=0d4431b2e4731ec58d6ee8d9f08ef4b3), [pyre](https://pyre-check.org/play?input=from%20typing%20import%20Any%2C%20Literal%2C%20assert_type%0A%0A%0Aclass%20CMBase%3A%0A%20%20%20%20def%20__enter__(self)%20-%3E%20None%3A%0A%20%20%20%20%20%20%20%20pass%0A%0A%0Aclass%20Suppress1(CMBase)%3A%0A%20%20%20%20def%20__exit__(self%2C%20exc_type%2C%20exc_value%2C%20traceback)%20-%3E%20bool%3A%0A%20%20%20%20%20%20%20%20return%20True%0A%0A%0Aclass%20Suppress2(CMBase)%3A%0A%20%20%20%20def%20__exit__(self%2C%20exc_type%2C%20exc_value%2C%20traceback)%20-%3E%20Literal%5BTrue%5D%3A%0A%20%20%20%20%20%20%20%20return%20True%0A%0A%0Aclass%20NoSuppress1(CMBase)%3A%0A%20%20%20%20def%20__exit__(self%2C%20exc_type%2C%20exc_value%2C%20traceback)%20-%3E%20None%3A%0A%20%20%20%20%20%20%20%20return%20None%0A%0A%0Adef%20suppress1(x%3A%20int%20%7C%20str)%20-%3E%20None%3A%0A%20%20%20%20if%20isinstance(x%2C%20int)%3A%0A%20%20%20%20%20%20%20%20with%20Suppress1()%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20raise%20ValueError%0A%20%20%20%20assert_type(x%2C%20int%20%7C%20str)%0A%0A%0Adef%20nosuppress1(x%3A%20int%20%7C%20str)%20-%3E%20None%3A%0A%20%20%20%20if%20isinstance(x%2C%20int)%3A%0A%20%20%20%20%20%20%20%20with%20NoSuppress1()%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20raise%20ValueError%0A%20%20%20%20assert_type(x%2C%20str)%0A) all pass.

Reply:
Ah, nice. Yeah, I was trying to think of other code flow patterns that might work better. I like your suggestion. I'll make that change.

Reply:
Hmm... Pyre shouldn't pass this test, because I know for a fact that we get this wrong :)

In Pyre, we incorrectly assume that context managers never swallow exceptions. This stems from a fundamental design flaw, where we construct the control flow graph in a phase before we have any type information, so we can't possibly inspect the return type annotation.

Reply:
Pyre passing in the playground seems to be an issue with the playground (I will investigate). When I try with a recent build in my dev environment, I see the expected failures for both suppress1 and suppress2.

```
Incompatible parameter type [6]: In call `assert_type`, for 1st positional argument, expected `Union[int, str]` but got `str`.
```

Reply:
@samwgoldman, hmm, I'm not sure how to square that with the test results. Pyre is emitting no errors (i.e. no `assert_type` failures) for this test, so it seems to be doing the right thing — unless there is some other bug that's causing it to not detect an `assert_type` mismatch. Any insights would be appreciated. If there's a better way to write this test, I'm happy to modify it.

Reply:
> ...I see the expected failures for both suppress1 and suppress2

Perhaps you're running an unreleased version of pyre where the behavior differs from the current released version? The [conformance tests](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) automatically pull the latest released version, which appears to be 0.9.19.

Reply:
It appears to be an issue with `assert_type()`; pyre allows both `assert_type(x, str)` and `assert_type(x, int | str)`.

Reply:
Ah, interesting. The conformance tests make extensive use of `assert_type`. If `assert_type` is not working correctly, that will affect many test results.

Reply:
Here's a variation that correctly finds the bug in pyre:

```
from typing import Any, Literal, assert_type


class CMBase:
    def __enter__(self) -> None:
        pass


class Suppress1(CMBase):
    def __exit__(self, exc_type, exc_value, traceback) -> bool:
        return True


class Suppress2(CMBase):
    def __exit__(self, exc_type, exc_value, traceback) -> Literal[True]:
        return True


class NoSuppress1(CMBase):
    def __exit__(self, exc_type, exc_value, traceback) -> None:
        return None


def suppress1(x: int | str) -> int | str:
    if isinstance(x, int):
        with Suppress1():
            raise ValueError
    return x + "x"  # E


def nosuppress1(x: int | str) -> str:
    if isinstance(x, int):
        with NoSuppress1():
            raise ValueError
    return x + "x"
```

https://pyre-check.org/play?input=from%20typing%20import%20Any%2C%20Literal%2C%20assert_type%0A%0A%0Aclass%20CMBase%3A%0A%20%20%20%20def%20__enter__(self)%20-%3E%20None%3A%0A%20%20%20%20%20%20%20%20pass%0A%0A%0Aclass%20Suppress1(CMBase)%3A%0A%20%20%20%20def%20__exit__(self%2C%20exc_type%2C%20exc_value%2C%20traceback)%20-%3E%20bool%3A%0A%20%20%20%20%20%20%20%20return%20True%0A%0A%0Aclass%20Suppress2(CMBase)%3A%0A%20%20%20%20def%20__exit__(self%2C%20exc_type%2C%20exc_value%2C%20traceback)%20-%3E%20Literal%5BTrue%5D%3A%0A%20%20%20%20%20%20%20%20return%20True%0A%0A%0Aclass%20NoSuppress1(CMBase)%3A%0A%20%20%20%20def%20__exit__(self%2C%20exc_type%2C%20exc_value%2C%20traceback)%20-%3E%20None%3A%0A%20%20%20%20%20%20%20%20return%20None%0A%0A%0Adef%20suppress1(x%3A%20int%20%7C%20str)%20-%3E%20int%20%7C%20str%3A%0A%20%20%20%20if%20isinstance(x%2C%20int)%3A%0A%20%20%20%20%20%20%20%20with%20Suppress1()%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20raise%20ValueError%0A%20%20%20%20return%20x%20%2B%20%22x%22%20%20%23%20E%0A%0A%0Adef%20nosuppress1(x%3A%20int%20%7C%20str)%20-%3E%20str%3A%0A%20%20%20%20if%20isinstance(x%2C%20int)%3A%0A%20%20%20%20%20%20%20%20with%20NoSuppress1()%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20raise%20ValueError%0A%20%20%20%20return%20x%20%2B%20%22x%22%0A

It produces no errors, so it fails because it doesn't produce the expected error. Mypy and pyright produce the correct error.

Reply:
Indeed, Steven fixed assert_type recently (a few weeks ago) in order to better support the conformance tests. I suspect that is the issue and we need to update the playground and re-record the conformance results.

Reply:
> Steven fixed assert_type recently

Ah, that's great to hear. If you publish a new version of pyre, the conformance test framework will automatically discover it and use it. Do you have a timeline for when you will publish the next version?